Allow Reference.refersTo to be inlined under AOT

hzongaro opened the pull request:
The implementations of `TR_J9SharedCacheVM::isGetImplInliningSupported` and `TR_J9SharedCacheVM::isGetImplAndRefersToInliningSupported` return false unconditionally, preventing inlining of `java/lang/ref/Reference.refersTo` and `java/lang/ref/Reference.getImpl` under AOT compilation.  This change removes the implementations of those methods in `TR_J9SharedCacheVM` to allow inlining to happen under AOT, with an additional test in `TR_J9VM::inlineNativeCall` of whether the 'referent' field's offset could not be retrieved - which can occur under AOT with no symbol validation manager.

hzongaro commented:
Irwin @dsouzai, may I ask you to review this change?

dsouzai commented on runtime/compiler/env/VMJ9.cpp:
```diff
@@ -7729,6 +7729,19 @@ TR_J9VM::inlineNativeCall(TR::Compilation * comp, TR::TreeTop * callNodeTreeTop,
+            int32_t offset =
+               getInstanceFieldOffset(ReferenceClass, REFERENCEFIELD, REFERENCEFIELDLEN, REFERENCERETURNTYPE, REFERENCERETURNTYPELEN, J9_RESOLVE_FLAG_INIT_CLASS);
+
+            // Guard against the possibility that the relevant field no longer exists
```
Under what circumstance could the field no longer exist?

hzongaro replied:
I will update the comment to make that more clear.

dsouzai replied:
Ah I see, it's not that it could change during runtime; I was more concerned from a relocation/validation point of view.

hzongaro replied:
Fixed in commit b03ac952857a54bfc3169e631a9d866990949f81

dsouzai commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -1544,8 +1544,6 @@ class TR_J9SharedCacheVM : public TR_J9VM
    virtual bool               sameClassLoaders(TR_OpaqueClassBlock *, TR_OpaqueClassBlock *);
    virtual bool               isUnloadAssumptionRequired(TR_OpaqueClassBlock *, TR_ResolvedMethod *);
    virtual bool               classHasBeenExtended(TR_OpaqueClassBlock *);
-   virtual bool               isGetImplInliningSupported();
-   virtual bool               isGetImplAndRefersToInliningSupported();
```
You should also remove these queries in `TR_J9SharedCacheServerVM` if it's being removed in `TR_J9SharedCacheVM`.

hzongaro replied:
Fixed in commit b03ac952857a54bfc3169e631a9d866990949f81

hzongaro commented:
Irwin @dsouzai, I've made the requested changes in commit b03ac952857a54bfc3169e631a9d866990949f81.  This is ready for a re-review.  Thanks!

dsouzai commented:
jenkins test sanity.functional all jdk17

dsouzai commented:
I'll let the tests complete, but I believe we need to increment JITServer minor version since there could be a miscommunication between the client and the server. I've already spoken to Henry about this offline.

hzongaro commented:
Irwin @dsouzai, I've updated the MINOR_NUMBER as we discussed.  Unfortunately, I had to resolve a merge conflict for the MINOR_NUMBER, so I've rebased and force pushed my commits with the latest versions.

There was an error in the [aarch64 build](https://openj9-jenkins.osuosl.org/job/Build_JDK17_aarch64_linux_Personal/455/) that appeared to be unrelated to my changes, but the other tests were clean.  I'll rerun the tests just to be safe.

hzongaro commented:
Jenkins test sanity.functional all jdk17

hzongaro commented:
Irwin @dsouzai, would you like me to squash down the two commits, so that the change to the MINOR_NUMBER is actually with the change to which it applied.

dsouzai commented:
Nah I think it's fine as is, it's all still part of the same PR anyway.
